src/main/java/com/google/research/bleth/utils/Queries.java

```diff
@@ -0,0 +1,88 @@
+import com.google.appengine.api.datastore.Query;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+/** A utility class providing method for db related operations, such as join and group by. */
+public class Queries {
+
+    /**
+     * Given a primary entity kind, a secondary entity kind, a foreign key and a filter, return a list of all entities
+     * of the secondary entity matching an entity of the primary entity kind.
+     *
+     * Equivalent SQL syntax:
+     *
+     * SELECT secondaryEntityKind.a_1, ... secondaryEntityKind.a_n
+     * FROM primaryEntityKind INNER JOIN secondaryEntityKind
+     * ON primaryEntityKind.__key__ = secondaryEntityKind.foreignKey
+     * WHERE primaryEntityFilter
+     *
+     * @param primaryEntityKind is the primary entity kind.
+     * @param secondaryEntityKind is the secondary entity kind.
+     * @param foreignKey is the join property of secondaryEntityKind (the join property of primaryEntityKind is the key).
+     * @param primaryEntityFilter is a simple or composed filter to apply on primaryEntityKind prior to the join operation (optional).
+     * @return a list of secondary entities matching the primary entities retrieved.
+     */
+    public static List<Entity> Join(String primaryEntityKind, String secondaryEntityKind,
+                                    String foreignKey, Optional<Query.Filter> primaryEntityFilter) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        List<Entity> result = new ArrayList<>();
+
+        // Retrieve all primaryEntity keys ordered by their key, filtered by primaryEntityFilter (if provided).
+        Query primaryMatchingFilterCondition = new Query(primaryEntityKind);
+        primaryEntityFilter.ifPresent(primaryMatchingFilterCondition::setFilter);
+        Iterator<String> primaryKeyIterator = datastore.prepare(primaryMatchingFilterCondition)
+                .asList(FetchOptions.Builder.withDefaults())
+                .stream()
+                .map(entity -> KeyFactory.keyToString(entity.getKey()))
+                .sorted()
+                .iterator();
+
+        // Retrieve all secondaryEntity ordered by their foreignKey.
+        Query secondary = new Query(secondaryEntityKind)
+                .addSort(foreignKey, Query.SortDirection.ASCENDING);
+        Iterator<Entity> secondaryEntityIterator = datastore.prepare(secondary).asIterator();
+
+        // If one of these queries is empty, return an empty list.
+        if (!(primaryKeyIterator.hasNext() && secondaryEntityIterator.hasNext())) {
+            return result;
+        }
+
+        // Iterate over two queries results and add secondary entities with a matching foreign key.
+        String primaryKey = primaryKeyIterator.next();
+        Entity secondaryEntity = secondaryEntityIterator.next();
+        String secondaryKey = String.valueOf(secondaryEntity.getProperty(foreignKey));
```

Comment:
Consider adding a helper method that does both of these steps together, so you don't need to duplicate them multiple times.

(Even better, introduce a second helper - with a similar name - that just wraps the `primaryKeyIterator.next()` call: then the code _here_ to advance both iterators would look the same.)

Reply:
I will leave it as is because of lines 74 and 83, as discussed.